### PR TITLE
ssfanist/Uniswap V3 support for reward token swap

### PR DESCRIPTION
**Pull request by defigods**
If you made a contract change, make sure to complete the checklist below before merging it in master.

Refer to our [documentation](https://github.com/OriginProtocol/security) for more details about contract security best practices.

Contract change checklist:
  - [ ] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [ ] Unit tests pass
  - [ ] Slither tests pass with no warning
  - [ ] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

**Comment by DanielVF** [comment]
The biggest thing I see that the choice to use the V3 or V2 interface needs to be a part of the configuration for the token. It should not be passed in from outside. It's okay then if there is only one uniswap address defined for a reward token, it's just how that address is called that will be changing.

File: contracts/contracts/harvest/Harvester.sol
Comment (DanielVF) [inline contracts/contracts/harvest/Harvester.sol:481]: Actually, lets change this to the 0.50% pool, since that new has funds in it.

File: contracts/contracts/harvest/Harvester.sol
Comment (tomlinton) [inline contracts/contracts/harvest/Harvester.sol:363]: We still need this comment. We can just update it to explain the v3 vs v2 functionality.

Comment (ssfanist) [reply contracts/contracts/harvest/Harvester.sol:363]: I just moved comments to below _swapV2 and _swapV3 functions.
But if you want, I'll put back to _swap function.
So should I remove the comments for _swapV2 and _swapV3 functions? or keep all 3 parts?

Comment (tomlinton) [reply contracts/contracts/harvest/Harvester.sol:363]: All 3 should have comments. They all do something different so they should be updated accordingly.

File: contracts/contracts/harvest/Harvester.sol
Comment (sparrowDom) [inline contracts/contracts/harvest/Harvester.sol:115]: nice catch thanks :)

File: contracts/contracts/harvest/Harvester.sol
Comment (sparrowDom) [inline contracts/contracts/harvest/Harvester.sol:159]: I would allow `_uniswapV2CompatibleAddr` to be zero address when the reward token is configured to be swapped using Uniswap V3.

A check that would make sense to me is `_uniswapV2CompatibleAddr ` != address(0) || _v3Fee > 0
meaning either UniswapV2 or Uniswap V3 should be configured. Maybe another check that both of them are not configured.

File: contracts/contracts/harvest/Harvester.sol
Comment (sparrowDom) [inline contracts/contracts/harvest/Harvester.sol:174]: Is this ok? I case token still wants to use UniswapV2 this is not going to work? I would: 
- token.safeApprove(0) && token.safeApprove(MAX)(as it was before this PR)  when the token configuration passed to the harvester is using non 0 `_uniswapV2CompatibleAddr `
- token.safeApprove(0) && token.safeApprove(MAX) when uniswap v3 is configured to be used.

File: contracts/deploy/037_dripper.js
Comment (sparrowDom) [inline contracts/deploy/037_dripper.js:111]: These deploys have already been ran and don't need this change

File: contracts/deploy/036_multiple_rewards_strategies.js
Comment (sparrowDom) [inline contracts/deploy/036_multiple_rewards_strategies.js:256]: as in the 037 migration file, I think this file doesn't need any changes, since it has already been executed on the mainnet

Comment (sparrowDom) [reply contracts/deploy/036_multiple_rewards_strategies.js:256]: You will probably need to create a fresh new deploy file (038...) in order to be able to upgrade the harvester.

**Review by sparrowDom — COMMENTED** [review]
Left a couple of comments inline

File: contracts/contracts/harvest/Harvester.sol
Comment (ssfanist) [reply contracts/contracts/harvest/Harvester.sol:159]: 💪

**Comment by ssfanist** [comment]
@DanielVF @tomlinton @sparrowDom @franckc Any more things to update here?

File: contracts/contracts/harvest/Harvester.sol
Comment (sparrowDom) [inline contracts/contracts/harvest/Harvester.sol:195]: If we want to be super clean we should call approve only when current _v3Fee == 0 (or not set) and the current one is not 0.

**Comment by sparrowDom** [comment]
Left 1 minor comment inline, and what is also required is a new completely separate `038` deploy script. All the other deploy scripts have already been executed on the mainnet. And each one executes only once.

Will let @DanielVF stamp the final seal of approval since he is sort of the code owner of this part.

**Comment by ssfanist** [comment]
I'm not sure what to do in the deploy script 38. Would you share with me? @tomlinton

**Comment by sparrowDom** [comment]
I would just copy all the things from script 036 and keep the part that upgrades the Harvester to the new version of the contract

File: contracts/contracts/harvest/Harvester.sol
Comment (DanielVF) [reply contracts/contracts/harvest/Harvester.sol:159]: I am impressed. Worked it out on paper to verify.

Basically this is doing an XOR so that one and only one of the two conditions is true. 

It's a little confusing to understand because one side is true when set, and the other side is false when set. I think it would be clearer if this used the normal XOR substitute "!=" for the middle comparison operator, and then did `(_v3Fee > 0)`. This way both sides of the check are the same kind, with true for this kind is set, and false for this kind is not set.

Thanks, I learned something today.

File: contracts/contracts/harvest/Harvester.sol
Comment (DanielVF) [reply contracts/contracts/harvest/Harvester.sol:174]: Echoing what Domen said, this whole token approve section should be two independent parts. One that handles uniswap v2 and one that handles uniswap v3. Each can just separately see if it needs to operate, and does not need to know about the other.

Currently this is also missing the second to unapprove uniswapV3.

File: contracts/test/vault/harvester.js
Comment (DanielVF) [inline contracts/test/vault/harvester.js:36]: The tests don't seem to test the uniswap v3 swapping? Or did I miss it?

**Review by DanielVF — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
